src/sinks/datadog/logs/sink.rs

```diff
@@ -158,14 +200,76 @@ impl From<io::Error> for RequestBuildError {
 
 struct LogRequestBuilder {
     default_api_key: Arc<str>,
-    encoding: EncodingConfigFixed<DatadogLogsJsonEncoding>,
+    encoding: EncodingConfigFixed<JsonEncoding>,
     compression: Compression,
 }
 
 impl RequestBuilder<(Option<Arc<str>>, Vec<Event>)> for LogRequestBuilder {
     type Metadata = (Arc<str>, usize, EventFinalizers, usize);
     type Events = Vec<Event>;
-    type Encoder = EncodingConfigFixed<DatadogLogsJsonEncoding>;
+    type Encoder = EncodingConfigFixed<JsonEncoding>;
+    type Payload = Bytes;
+    type Request = LogApiRequest;
+    type Error = RequestBuildError;
+
+    fn compression(&self) -> Compression {
+        self.compression
+    }
+
+    fn encoder(&self) -> &Self::Encoder {
+        &self.encoding
+    }
+
+    fn split_input(&self, input: (Option<Arc<str>>, Vec<Event>)) -> (Self::Metadata, Self::Events) {
+        let (api_key, mut events) = input;
+        let events_len = events.len();
+        let finalizers = events.take_finalizers();
+        let events_byte_size = events.size_of();
+
+        let api_key = api_key.unwrap_or_else(|| Arc::clone(&self.default_api_key));
+        ((api_key, events_len, finalizers, events_byte_size), events)
+    }
+
+    fn encode_events(&self, events: Self::Events) -> Result<Self::Payload, Self::Error> {
+        // We need to first serialize the payload separately so that we can figure out how big it is
+        // before compression.  The Datadog Logs API has a limit on uncompressed data, so we can't
+        // use the default implementation of this method.
+        let mut buf = Vec::new();
+        let n = self.encoder().encode_input(events, &mut buf)?;
+        if n > MAX_PAYLOAD_BYTES {
+            return Err(RequestBuildError::PayloadTooBig);
+        }
+
+        // Now just compress it like normal.
+        let mut compressor = Compressor::from(self.compression);
+        let _ = compressor.write_all(&buf)?;
+
+        Ok(compressor.into_inner().freeze())
+    }
+
+    fn build_request(&self, metadata: Self::Metadata, payload: Self::Payload) -> Self::Request {
+        let (api_key, batch_size, finalizers, events_byte_size) = metadata;
+        LogApiRequest {
+            batch_size,
+            api_key,
+            compression: self.compression,
+            body: payload,
+            finalizers,
+            events_byte_size,
+        }
+    }
+}
+
+struct SemanticLogRequestBuilder {
+    default_api_key: Arc<str>,
+    encoding: EncodingConfigFixed<SemanticJsonEncoding>,
+    compression: Compression,
+}
```

Comment:
We should remove this duplication by integrating this into the `LogRequestBuilder` as a generic type.

Reply:
Follow-up. I'll create an issue before merging 👍